osclib/core.py

```diff
@@ -414,26 +414,43 @@ def attribute_value_save(apiurl, project, name, value, namespace='OSRT', package
     url = makeurl(apiurl, list(filter(None, ['source', project, package, '_attribute'])))
     http_POST(url, data=ET.tostring(root))
 
+
 def attribute_value_delete(apiurl, project, name, namespace='OSRT', package=None):
     http_DELETE(makeurl(
         apiurl, list(filter(None, ['source', project, package, '_attribute', namespace + ':' + name]))))
 
-@memoize(session=True)
+
 def repository_path_expand(apiurl, project, repo):
     """Recursively list underlying projects."""
+    visited_repos = set(f"{project}:{repo}")
+    return _recursive_repository_path_expand(apiurl, project, repo, visited_repos)
+
+
+@memoize(session=True)
+def _recursive_repository_path_expand(apiurl, project, repo, visited_repos):
+    """internal helper for avoiding recursing while building recursive list underlying projects."""
     repos = [[project, repo]]
+    visited_repos.add(f"{project}:{repo}")
     meta = ET.fromstringlist(show_project_meta(apiurl, project))
     paths = meta.findall('.//repository[@name="{}"]/path'.format(repo))
 
     # The listed paths are taken as-is, except for the last one...
     for path in paths[:-1]:
-        repos += [[path.get('project', project), path.get('repository')]]
+        p_project = path.get('project', project)
+        p_repository = path.get('repository')
+        visited_repos.add(f"{p_project}:{p_repository}")
+        repos += [[p_project, p_repository]]
 
     # ...which is expanded recursively
     if len(paths) > 0:
-        repos += repository_path_expand(apiurl, paths[-1].get('project', project), paths[-1].get('repository'))
+        p_project = paths[-1].get('project', project)
+        p_repository = paths[-1].get('repository')
+        if f"{p_project}:{p_repository}" not in visited_repos:
+            visited_repos.add(f"{p_project}:{p_repository}")
+            repos += _recursive_repository_path_expand(apiurl, p_project, p_repository, visited_repos)
```

Comment:
There's already `repos` which could be passed instead of `visited_repos`

Reply:
no, we want to add the one we recursed on.